armi/physics/neutronics/tests/test_crossSectionManager.py

```diff
@@ -104,21 +105,198 @@ def setUp(self):
+
+    def test_createRepresentativeBlockDissimilar(self):
+        """
+        Test creation of a representative block from a collection with dissimilar blocks
+        """
+
+        uniqueBlock = test_blocks.loadTestBlock()
+        uniqueBlock.p.percentBu = 50.0
+        fpFactory = test_lumpedFissionProduct.getDummyLFPFile()
+        uniqueBlock.setLumpedFissionProducts(fpFactory.createLFPsFromFile())
+        uniqueBlock.setNumberDensity("U235", 2.0)
+        uniqueBlock.p.gasReleaseFraction = 1.0
+        for c in uniqueBlock:
+            if c.hasFlags(Flags.FUEL):
+                c.temperatureInC = 600.0
+            elif c.hasFlags([Flags.CLAD, Flags.DUCT, Flags.WIRE]):
+                c.temperatureInC = 500.0
+            elif c.hasFlags([Flags.BOND, Flags.COOLANT, Flags.INTERCOOLANT]):
+                c.temperatureInC = 400.0
+        self.bc.append(uniqueBlock)
+
+        with mockRunLogs.BufferLog() as mock:
+            avgB = self.bc.createRepresentativeBlock()
+            self.assertIn(
+                "Non-matching block in AverageBlockCollection", mock.getStdout()
+            )
+
+        self.assertNotIn(avgB, self.bc)
+        # (0 + 1 + 2 + 3 + 4 + 2) / 6.0 = 12/6 = 2.0
+        self.assertAlmostEqual(avgB.getNumberDensity("U235"), 2.0)
+        # (0 + 1/4 + 2/4 + 3/4 + 4/4) / 5 * 100.0 = 50.0
+        self.assertAlmostEqual(avgB.p.percentBu, 50.0)
+
+        # U35 has different average temperature because blocks have different U235 content
+        newBc = AverageBlockCollection(
+            self.blockList[0].r.blueprints.allNuclidesInProblem
+        )
+        newBc.append(avgB)
+        newBc.calcAvgNuclideTemperatures()
+        # temps expected to be proportional to volume-fraction weighted temperature
+        # this is a non-physical result, but it demonstrates a problem that exists in the code
+        # when dissimilar blocks are put together in a BlockCollection
+        structureVolume = sum(
+            c.getVolume()
+            for c in avgB.getComponents([Flags.CLAD, Flags.DUCT, Flags.WIRE])
+        )
+        fuelVolume = avgB.getComponent(Flags.FUEL).getVolume()
+        coolantVolume = sum(
+            c.getVolume()
+            for c in avgB.getComponents([Flags.BOND, Flags.COOLANT, Flags.INTERCOOLANT])
+        )
+        expectedIronTemp = (structureVolume * 500.0 + fuelVolume * 600.0) / (
+            structureVolume + fuelVolume
+        )
+        expectedSodiumTemp = (coolantVolume * 400.0 + fuelVolume * 600.0) / (
+            coolantVolume + fuelVolume
+        )
+        self.assertAlmostEqual(newBc.avgNucTemperatures["U235"], 600.0)
+        self.assertAlmostEqual(newBc.avgNucTemperatures["FE56"], expectedIronTemp)
+        self.assertAlmostEqual(newBc.avgNucTemperatures["NA23"], expectedSodiumTemp)
+
+
+class TestComponentAveraging(unittest.TestCase):
+
+    def setUp(self):
+        fpFactory = test_lumpedFissionProduct.getDummyLFPFile()
+        self.blockList = makeBlocks(3)
```

Comment:
You can improve performance a little by moving `makeBlocks` to a `setUpClass` method. `makeBlocks` calls the `loadTestReactor` method. From the looks of it, that only needs to be called once here.

Reply:
Got it down from 2.5 s to 1.66 s for the 3 new tests!

Reply:
I used `setUpClass` for the other test classes also.